### ISSUE TITLE
Parse error, probably to do with (assumed, but not actually present) quasi quotation

### ISSUE THREAD

**Issue by jvoigtlaender** [ndmitchell/hlint#1038]
For this file:
```haskell
x :: [Int]
x = [e| e <- [] ]
```
I get:
```
> hlint --version
HLint v3.1.4, (C) Neil Mitchell 2006-2020
> hlint.exe --language=NoQuasiQuotes .\Test.hs
Test.hs:2:11-12: Error: Parse error: on input `<-'
Perhaps this statement should be within a 'do' block?
Found:
    x :: [Int]
  > x = [e| e <- [] ]

1 hint
```
The error goes away if I add a space next to the first `e` or if I replace the variable `e` by `y`, say.

**Comment by ndmitchell** [comment]
The issue is `TemplateHaskell`, see https://downloads.haskell.org/~ghc/latest/docs/html/users_guide/glasgow_exts.html#th-syntax, which enables `[e| ... |]`. You can ignore it with `-XNoTemplateHaskell`. We actually have QuasiQuotes off by default because it breaks `[foo| x]` which happens enough, while `TemplateHaskell` only steals `e`, `d` and `t`. I'd really like to enable Template Haskell but not those prefix ones, but that's not possible :(.

**Comment by jvoigtlaender** [comment]
I had already tried it with `--language=NoTemplateHaskell`, with same result. And also `-XNoTemplateHaskell` doesn't seem to have the effect (for hlint, other than for ghc) to really turn this feature off:
```
> cat .\Test.hs
x :: [Int]
x = [e| e <- [] ]

> hlint -XNoTemplateHaskell .\Test.hs
Test.hs:2:11-12: Error: Parse error: on input `<-'
Perhaps this statement should be within a 'do' block?
Found:
    x :: [Int]
  > x = [e| e <- [] ]

1 hint

> hlint --language=NoTemplateHaskell .\Test.hs
Test.hs:2:11-12: Error: Parse error: on input `<-'
Perhaps this statement should be within a 'do' block?
Found:
    x :: [Int]
  > x = [e| e <- [] ]

1 hint
```

But *actually*, I just found that this works fine:
```
> hlint --language=NoTemplateHaskellQuotes .\Test.hs
Test.hs:2:5-17: Suggestion: Redundant list comprehension
Found:
  [e | e <- []]
Perhaps:
  []

1 hint
``` 

But probably your point still stands that you can't have `TemplateHaskell` and also `NoTemplateHaskellQuotes`.

**Comment by ndmitchell** [comment]
Hmm, it's a little frustrating that `-XNoTemplateHaskell` doesn't disable `TemplateHaskellQuotes` too - unfortunately the implication only works in one way.

**Comment by jvoigtlaender** [comment]
Yeah, though I must say I am also a bit confused now. So `TemplateHaskell` "implies" `TemplateHaskellQuotes`. But does that mean that if one specifies both `TemplateHaskell` and `NoTemplateHaskellQuotes`, then the latter is ignored and `TemplateHaskellQuotes` is on anyway?

If not, then maybe the solution to your intention:
> I'd really like to enable Template Haskell but not those prefix ones, but that's not possible :(.

would be to overall have as default settings for hlint in this regard the combination of the following three settings?
- `NoQuasiQuotes`
- `TemplateHaskell`
- `NoTemplateHaskellQuotes`

**Comment by ndmitchell** [comment]
I'd like to allow `[|` as TemplateHaskell, but treat `[e|` as a list comp. There isn't any extension that controls that - `TemplateHaskellQuotes` controls both, so I have to pick which seems worse. My guess is since `[e |` would be the usual form that `TemplateHaskellQuotes` should be turned on by default.

To make it a bit more pleasant, I've added code so `-XNoTemplateHaskell` implies `-XNoTemplateHaskellQuotes`, which should make it easier to figure out what to disable.

**Comment by jvoigtlaender** [comment]
Okay, so I guess this issue can be closed, then?

**Comment by ndmitchell** [comment]
Yep, sounds like the best thing to do.
